### PR TITLE
Allow setting RootNamespace in ResourceLocationAttribute

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This fixes the problem of us lacking a RootNamespace at runtime. I've made it a separate attribute so that when we create it using a task in the future people won't have to worry about the task interfeering with their `ResourceLocationAttribute`.

### CODE REVIEW

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+using System;
+
+namespace Microsoft.Extensions.Localization
+{
+    /// <summary>
+    /// Provides the RootNamespace of an Assembly.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false, Inherited = false)]
+    public class RootNamespaceAttribute : Attribute
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: I'm not sure if this is a correct place for such property!!

**Comment by hishamco** [comment]
IMHO adding this to `LocalizationOptions` will be better
```csharp
services.AddLocalization(options => {
     options.ResourcesPath = "My/Resources";
     options.RootNamespace = "Alternate Namespace";
});
```

**Comment by ryanbrandenburg** [comment]
@hishamco that doesn't solve the problem for Class libraries, and once we get the task which automatically adds the attribute to your assembly at compile time running, it would only be an extra place you have to update the namespace if you change it.

**Comment by rynowak** [comment]
> IMHO adding this to LocalizationOptions will be better

If it were that simple, I'd totally agree with you 😆 

The root namespace isn't global, it's per assembly

**Comment by ryanbrandenburg** [comment]
Because if an assembly already manually creates the `ResourceLocationAttribute` (which you can only have one of) we don't want them to have problems when we start auto-generating the attribute for RootNamespace.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 2]
```diff
@@ -68,7 +68,7 @@ protected virtual string GetResourcePrefix(TypeInfo typeInfo)
                 throw new ArgumentNullException(nameof(typeInfo));
             }
 
-            return GetResourcePrefix(typeInfo, new AssemblyName(typeInfo.Assembly.FullName).Name, _resourcesRelativePath);
+            return GetResourcePrefix(typeInfo, GetRootNamespace(typeInfo.Assembly), GetResourcePath(typeInfo.Assembly));
```
Comment (kichalla) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Why the change `GetResourcePath(typeInfo.Assembly)`? This is a change in behavior?

Comment (dougbu) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Also, why bother changing the `baseNamespace` argument in a call when you're changing the called overload to ignore `baseNamespace`?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 3]
```diff
@@ -93,10 +93,11 @@ protected virtual string GetResourcePrefix(TypeInfo typeInfo, string baseNamespa
             {
                 throw new ArgumentNullException(nameof(baseNamespace));
             }
+            var assemblyName = new AssemblyName(typeInfo.Assembly.FullName).Name;
 
             return string.IsNullOrEmpty(resourcesRelativePath)
                 ? typeInfo.FullName
-                : baseNamespace + "." + resourcesRelativePath + TrimPrefix(typeInfo.FullName, baseNamespace + ".");
+                : baseNamespace + "." + resourcesRelativePath + TrimPrefix(typeInfo.FullName, assemblyName + ".");
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Why does this method now ignore `baseNamespace` (except for the `null` or empty check)?

File: test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
+    <RootNamespace>Alternate.Namespace</RootNamespace>
```
Comment (dougbu) [inline test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj]: Is this change necessary?

`$(RootNamespace)` normally only controls the namespace of generated files such as `Resources.Designer.cs` or new files in the project. Since Localization uses only the resources and not the generated class, what matters here?

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        /// <summary>
+        /// Creates a new <see cref="RootNamespaceAttribute"/>.
+        /// </summary>
+        /// <param name="rootNamespace">The RootNamespace for this Assembly.</param>
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: This comment doesn't mean much. Describe what the root namespace controls.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 6]
```diff
@@ -217,6 +218,20 @@ protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembl
             return assembly.GetCustomAttribute<ResourceLocationAttribute>();
         }
 
+        protected virtual RootNamespaceAttribute GetRootNamespaceAttribute(Assembly assembly)
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: In what scenario would users (not our tests) override this method? Leaning toward `internal virtual` instead of exposing this publicly.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 2 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: I spoke with @dougbu in person and we agreed against his comments. @kichalla we changed to use GetResourcePath here because it's correct. Otherwise we would be ignoring the ResourceLocationAttribute in some cases.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 3 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Spoke with @dougbu in person and we now agree this is correct.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 6 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: I chose `protected virutal` for the same reasons I did for `GetResourceLocationAttribute` just above this. The compromise we reached then was to include a warning against using it in the remarks, which I'll add in my reaction commit.

File: test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj (hunk 4 above)
Comment (ryanbrandenburg) [reply test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj]: This change is necessary. To prove it remove `[assembly: RootNamespace("Alternate.Namespace")]` from `ResourcesClassLibraryWithAttribute\AssemblyInfo.cs` in this PR and watch resources not be found. This is because dotnet uses RootNamespace as part of the name when automatically adding resources to the assembly, but `RootNamespace` is not available at runtime.

**Comment by ryanbrandenburg** [comment]
🆙:date:

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 7]
```diff
@@ -94,9 +94,17 @@ protected virtual string GetResourcePrefix(TypeInfo typeInfo, string baseNamespa
                 throw new ArgumentNullException(nameof(baseNamespace));
             }
 
-            return string.IsNullOrEmpty(resourcesRelativePath)
-                ? typeInfo.FullName
-                : baseNamespace + "." + resourcesRelativePath + TrimPrefix(typeInfo.FullName, baseNamespace + ".");
+            if (string.IsNullOrEmpty(resourcesRelativePath))
+            {
+                return typeInfo.FullName;
```
Comment (kichalla) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Should this consider the `baseNamespace` here as one could have a RootNamespaceAttribute but no resources relative path?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 8]
```diff
@@ -94,9 +94,17 @@ protected virtual string GetResourcePrefix(TypeInfo typeInfo, string baseNamespa
...
+            }
+            else
+            {
+                // This expectation is defined by dotnet's automatic resource storage.
+                // We have to conform to "{RootNamespace}.{ResourceLocation}.{FullTypeName - AssemblyName}".
+                var assemblyName = new AssemblyName(typeInfo.Assembly.FullName).Name;
+                return baseNamespace + "." + resourcesRelativePath + TrimPrefix(typeInfo.FullName, assemblyName + ".");
```
Comment (kichalla) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Trim the `baseNamespace`, `resourcesRelativePath` for `.` here and any other methods that users(as they are open for extension) can passing in value.

File: test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj (hunk 4 above)
Comment (dougbu) [reply test/ResourcesClassLibraryWithAttribute/ResourcesClassLibraryWithAttribute.csproj]: Got it

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 7 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: To my knowledge if you don't have a `resourcesRelativePath` the the resource name is just the full type name.

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    /// Provides the RootNamespace of an Assembly.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false, Inherited = false)]
+    public class RootNamespaceAttribute : Attribute
+    {
+        /// <summary>
+        /// Creates a new <see cref="RootNamespaceAttribute"/>.
+        /// </summary>
+        /// <param name="rootNamespace">The RootNamespace for this Assembly. The RootNamespace of the assembly is used by Localization to
+        /// determine the resource name to look for when RootNamespace differs from the AssemblyName.</param>
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: These comments will be much less visible than those for the class. (One of the general `Attribute` oddities.) Suggest moving the second sentence up into the class' `<summary>`.

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            {
+                throw new ArgumentNullException(nameof(rootNamespace));
+            }
+
+            RootNamespace = rootNamespace;
+        }
+
+        /// <summary>
+        /// The RootNamespace of this Assembly. The RootNamespace of the assembly is used by Localization to
+        /// determine the resource name to look for when RootNamespace differs from the AssemblyName.
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: 🥇

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 11]
```diff
@@ -111,6 +122,34 @@ public void Create_FromType_ReturnsNewResultForDifferentType()
             Assert.NotSame(result1, result2);
         }
 
+        [Fact]
+        public void Create_ResourceLocationAttribute_UsesRootNamespace()
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Need a test enshrining the has `RootNamespaceAttribute` / doesn't have `ResourceLocationAttribute` behaviour.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 8 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Good point that we need to start doing that, but we should do so holistically for all such inputs at once and add tests, so I've created aspnet/Home#2792 since it seems like a larger item.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs (hunk 11 above)
Comment (ryanbrandenburg) [reply test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: I don't know what you mean by this, could you expand?

Comment (dougbu) [reply test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: I mean this test illustrates what happens when both attributes are in play. From your comments on the product code, I understand that `[assembly: RootNamespace(...)]` is a no-op when there's no `ResourceLocationAttribute` in play. Need a test demonstrating that.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 7 above)
Comment (kichalla) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Why not write a test for this case? or you can find out from a test app in VS

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs (hunk 5 above)
Comment (dougbu) [reply src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: Done above. Thanks.

File: src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            if (string.IsNullOrEmpty(rootNamespace))
+            {
+                throw new ArgumentNullException(nameof(rootNamespace));
+            }
+
+            RootNamespace = rootNamespace;
+        }
+
+        /// <summary>
+        /// The RootNamespace of this Assembly.
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/RootNamespaceAttribute.cs]: Suggest leaving the explanation here too.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 7 above)
Comment (dougbu) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: `Create_ResourceLocationAttribute_RootNamespaceIgnoredWhenNoLocation` has an odd name but is that test.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 13]
```diff
@@ -217,6 +225,24 @@ protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembl
...
+            return assembly.GetCustomAttribute<RootNamespaceAttribute>();
+        }
+
+        private string GetRootNamespace(Assembly assembly)
+        {
+            var rootNamespaceAttribute = GetRootNamespaceAttribute(assembly);
+
+            return rootNamespaceAttribute?.RootNamespace == null
+                ? new AssemblyName(assembly.FullName).Name
+                : rootNamespaceAttribute.RootNamespace;
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Completely up to you: `rootNamespaceAttribute?.RootNamespace ?? new AssemblyName(assembly.FullName).Name`

Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: 👍

**Comment by ryanbrandenburg** [comment]
@kichalla do you approve?
